llvm/test/CodeGen/AIE/alias-analysis-pointer-desambiguation.ll

```diff
@@ -0,0 +1,557 @@
+; ENABLED-AIE-AA-DIS:  NoAlias:      i8* %10, i8* %ptr_store
+; ENABLED-AIE-AA-DIS:  NoAlias:      i8* %10, i8* %3
+; ENABLED-AIE-AA-DIS:  NoAlias:      i8* %10, i8 addrspace(3)* %ptrcast 
+; ENABLED-AIE-AA-DIS:  NoAlias:      i8* %14, i8* %ptr_load
+; ENABLED-AIE-AA-DIS:  NoAlias:      i8* %14, i8* %ptr_store
+; ENABLED-AIE-AA-DIS:  NoAlias:      i8* %14, i8* %3
+; ENABLED-AIE-AA-DIS:  NoAlias:      i8* %14, i8 addrspace(3)* %ptrcast 
+; ENABLED-AIE-AA-DIS:  MayAlias:     i8* %10, i8* %14
+
+; DISABLED-AIE-AA-DES-LABEL: Function: with_addrspace_cast
+; DISABLED-AIE-AA-DES:  MayAlias:      i8* %ptr_load, i8* %ptr_store
+; DISABLED-AIE-AA-DES:  MayAlias:      i8* %3, i8* %ptr_load
+; DISABLED-AIE-AA-DES:  MayAlias:      i8* %3, i8* %ptr_store
+; DISABLED-AIE-AA-DES:  MayAlias:      i8* %ptr_load, i8 addrspace(3)* %ptrcast
+; DISABLED-AIE-AA-DES:  MayAlias:      i8* %ptr_store, i8 addrspace(3)* %ptrcast
+; DISABLED-AIE-AA-DES:  MayAlias:      i8* %3, i8 addrspace(3)* %ptrcast
+; DISABLED-AIE-AA-DES:  MayAlias:      i8* %10, i8* %ptr_load
+; DISABLED-AIE-AA-DES:  MayAlias:      i8* %10, i8* %ptr_store
+; DISABLED-AIE-AA-DES:  MayAlias:      i8* %10, i8* %3
+; DISABLED-AIE-AA-DES:  MayAlias:      i8* %10, i8 addrspace(3)* %ptrcast
+; DISABLED-AIE-AA-DES:  MayAlias:      i8* %14, i8* %ptr_load
+; DISABLED-AIE-AA-DES:  MayAlias:      i8* %14, i8* %ptr_store
+; DISABLED-AIE-AA-DES:  MayAlias:      i8* %14, i8* %3
+; DISABLED-AIE-AA-DES:  MayAlias:      i8* %14, i8 addrspace(3)* %ptrcast
+; DISABLED-AIE-AA-DES:  MayAlias:      i8* %10, i8* %14
+
+; Function Attrs: nofree norecurse nounwind mustprogress
+define dso_local i32 @with_addrspace_cast(ptr %a, i8 %b, i32 %c) local_unnamed_addr #0 {
+entry:
+  br label %for.body
+
+for.body:                                         ; preds = %entry, %for.body
+  %counter_load = phi i20 [ %9, %for.body ], [ 0, %entry ]
+  %counter_store = phi i20 [ %13, %for.body ], [ 0, %entry ]
+  %i.08 = phi i32 [ %inc, %for.body ], [ 0, %entry ]
+  %ptr_load = phi ptr [ %10, %for.body ], [ %a, %entry ]
+  %ptr_store = phi ptr [ %14, %for.body ], [ %a, %entry ]
+  load i8, ptr %ptr_load
+  store i8 %b, ptr %ptr_store
+
+  %1 = tail call { ptr, i20 } @llvm.aie2.add.2d(ptr %ptr_load, i20 0, i20 1, i20 2, i20 %counter_load)
+  %2 = extractvalue { ptr, i20 } %1, 1
+  %3 = extractvalue { ptr, i20 } %1, 0
+  load i8, ptr %3
+  
+  %5 = tail call { ptr, i20 } @llvm.aie2.add.2d(ptr %ptr_store, i20 0, i20 1, i20 2, i20 %counter_store)
+  %6 = extractvalue { ptr, i20 } %5, 1
+  %7 = extractvalue { ptr, i20 } %5, 0
+
+  %ptrcast = addrspacecast ptr %7 to ptr addrspace(3)
+
+  store i8 %b, ptr addrspace(3) %ptrcast
+
+  %8 = tail call { ptr, i20 } @llvm.aie2.add.2d(ptr %3, i20 0, i20 1, i20 2, i20 %2)
+  %9 = extractvalue { ptr, i20 } %8, 1
+  %10 = extractvalue { ptr, i20 } %8, 0
+  load i8, ptr %10
+
+  %12 = tail call { ptr, i20 } @llvm.aie2.add.2d(ptr %7, i20 0, i20 1, i20 2, i20 %6)
```

Comment:
Is there any way to get the alias info for `%7`? Maybe by inserting a store to `%7` on top of the one for `%ptrcast`?

I'm suspecting that the alias info will be wrong for `%7`, because in the code below, I think that `LastUpdate` will point to `%ptrcast` after visiting the instrinsic, which is different from `%7`, so the update of `LastCalls` will not trigger.
```
%ptrcast = addrspacecast ptr %7 to ptr addrspace(3)
%12 = tail call { ptr, i20 } @llvm.aie2.add.2d(ptr %7, i20 0, i20 1, i20 2, i20 %6)
```

Reply:
The cast situation is handled by the `@with_addrspace_cast_phi` test.